### PR TITLE
refactor(storage): add RestClient runtime injection and testing

### PR DESCRIPTION
**Pull request by scotthart**
This PR is the beginning of refactoring of switching GCS over to use the new REST library. Currently RestClient redirects all calls to the existing implementation, CurlClient, and RestClient is only injected at runtime in lieu of CurlClient when the GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT env var is present.

It also adds two new testing builds that define the env var.

In the future, additional PRs will begin to reimplement RestClients methods to use the REST library.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778)

### CODE REVIEW

File: ci/cloudbuild/builds/cmake-gcs-rest-install.sh [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+# parent directories must also exist.
+mapfile -t actual_dirs < <(env -C "${INSTALL_PREFIX}" find -type d)
+# The micro generator populates `ci/etc/expected_install_directories` with the
+# expected directories generated via `make install`. This generated list does
+# not include:
+#   - hand-crafted libraries
+#   - directories with pure protos (without gRPC code)
+# We maintain those here as a hard-coded list.
+mapfile -t expected_dirs < <(cat ci/etc/expected_install_directories)
+expected_dirs+=(
```
Comment (coryan) [inline ci/cloudbuild/builds/cmake-gcs-rest-install.sh:56]: I assume this list must match the list in `cmake-install.sh`?  Can we refactor it to a text file and load it from both scripts?

File: google/cloud/storage/google_cloud_cpp_storage.cmake [hunk 2]
```diff
@@ -235,6 +237,7 @@ target_link_libraries(
            absl::time
            absl::variant
            google-cloud-cpp::common
+           google-cloud-cpp::rest-internal
```
Comment (coryan) [inline google/cloud/storage/google_cloud_cpp_storage.cmake:240]: This means that any time `storage` is enabled (via `GOOGLE_CLOUD_CPP_ENABLE`) we also need to set `GOOGLE_CLOUD_CPP_ENABLE_REST`:

https://github.com/googleapis/google-cloud-cpp/blob/8a2298482d4ecbdcbc3406afa65c3d4f44b1d749/google/cloud/CMakeLists.txt#L47


That SGTM.  I wonder if we should detect an incompatible setting. For example, we could issue a fatal error or at least a warning if somebody explicitly sets `-DGOOGLE_CLOUD_CPP_ENABLE=storage -DGOOGLE_CLOUD_CPP_ENABLE_REST=OFF` in their CMake configuration.

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 14 of 14 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778)* status: all files reviewed, 3 unresolved discussions (waiting on @scotthart)

___
*[`ci/cloudbuild/builds/cmake-gcs-rest-install.sh` line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0Cajw4oV1k-B-VVSdg-r1-1:-N0h2DIe-sGaPDq4jwUY:b22bvyi) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/35cd4747ae3df21b989729f7a157098b2d120bc0/ci/cloudbuild/builds/cmake-gcs-rest-install.sh#L1)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I wanted to make sure that the rest library is being installed correctly in combination with storage.
</blockquote></details>

But that would be tested by `cmake-install.sh` right?  It does install `storage` and therefore it must install `rest-internal`? You could create a build to verify that `rest-internal` installs correctly even if `storage` is disabled, but that is such a rare use case that we probably can avoid the additional test.
___
*[`google/cloud/storage/client.cc` line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0h3gDREZVBZTErL1WH:-N0h3gDREZVBZTErL1WI:bs1vufx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/65cfcaca8ad6a808c6cd8a67e0f61680b5b5f51f/google/cloud/storage/client.cc#L60)):*
> ```cpp
>     Options const& opts) {
>   if (google::cloud::internal::GetEnv(
>           "GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT")
> ```

Should this be `HAVE_REST_CLIENT` or should this be `USE_REST_CLIENT`?

Also, do you think a `storage_experimental::UseRestClient(...)` option would be useful?
___
*[`google/cloud/storage/client_test.cc` line 168 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0h2oHE342I5_resYqg:-N0h2oHE342I5_resYqh:b-akuuhb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/65cfcaca8ad6a808c6cd8a67e0f61680b5b5f51f/google/cloud/storage/client_test.cc#L168)):*
> ```cpp
>   ASSERT_TRUE(retry != nullptr);
> 
>   if (google::cloud::internal::GetEnv(
> ```

Hmmm now that I think about it, typically we write a couple of different tests for these, e.g.:

```cc
TEST_F(ClientTest, WithCurlBackend) {
  ScopedEnvironment env("GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT", absl::nullopt);
  auto tested = Client(Options{}...);
....
  ASSERT_THAT(dynamic_cast<internal::CurlClient*>(retry->client().get()), NotNull());
}

TEST_F(ClientTest, WithRestBackend) {
  ScopedEnvironment env("GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT", "yes");
  auto tested = Client(Options{}...);
....
  ASSERT_THAT(dynamic_cast<internal::RestClient*>(retry->client().get()), NotNull());
}

```

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778)* status: 14 of 15 files reviewed, 3 unresolved discussions (waiting on @coryan)

___
*[`ci/cloudbuild/builds/cmake-gcs-rest-install.sh` line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0Cajw4oV1k-B-VVSdg-r1-1:-N0lAnbu6QcAKTp0ydJ7:b1o8210) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/35cd4747ae3df21b989729f7a157098b2d120bc0/ci/cloudbuild/builds/cmake-gcs-rest-install.sh#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

But that would be tested by `cmake-install.sh` right?  It does install `storage` and therefore it must install `rest-internal`? You could create a build to verify that `rest-internal` installs correctly even if `storage` is disabled, but that is such a rare use case that we probably can avoid the additional test.
</blockquote></details>

This build also defines the env var to inject RestClient in lieu of CurlClient via cmake. I wanted to have both bazel and cmake builds that exercised this path. I chose a cmake install build to also make sure that the libraries were installed correctly for when we wanted to use the Rest library.
___
*[`google/cloud/storage/client.cc` line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0h3gDREZVBZTErL1WH:-N0lBfYRBQTzNpgurZNM:b-c53456) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/65cfcaca8ad6a808c6cd8a67e0f61680b5b5f51f/google/cloud/storage/client.cc#L60)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Should this be `HAVE_REST_CLIENT` or should this be `USE_REST_CLIENT`?

Also, do you think a `storage_experimental::UseRestClient(...)` option would be useful?
</blockquote></details>

The naming was chosento be consistent with HAVE_GRPC.

I think the option might be useful if we had a fully functional RestClient in storage and we wanted to give users the option at compile time to use on or the other. For this incremental refactoring effort, I think we want a runtime switch for which client to use.
___
*[`google/cloud/storage/client_test.cc` line 168 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0h2oHE342I5_resYqg:-N0lHtnV9arafGxqfOSU:b-foj8a9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/65cfcaca8ad6a808c6cd8a67e0f61680b5b5f51f/google/cloud/storage/client_test.cc#L168)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Hmmm now that I think about it, typically we write a couple of different tests for these, e.g.:

```cc
TEST_F(ClientTest, WithCurlBackend) {
  ScopedEnvironment env("GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT", absl::nullopt);
  auto tested = Client(Options{}...);
....
  ASSERT_THAT(dynamic_cast<internal::CurlClient*>(retry->client().get()), NotNull());
}

TEST_F(ClientTest, WithRestBackend) {
  ScopedEnvironment env("GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT", "yes");
  auto tested = Client(Options{}...);
....
  ASSERT_THAT(dynamic_cast<internal::RestClient*>(retry->client().get()), NotNull());
}

```

</blockquote></details>

Tests changed accordingly.

**Review by coryan — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778)* status: all files reviewed, 2 unresolved discussions (waiting on @scotthart)

___
*[`ci/cloudbuild/builds/cmake-gcs-rest-install.sh` line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0Cajw4oV1k-B-VVSdg-r1-1:-N0lX9TD9A03ssL9EaVK:bbr6m71) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/35cd4747ae3df21b989729f7a157098b2d120bc0/ci/cloudbuild/builds/cmake-gcs-rest-install.sh#L1)):*
> This build also defines the env var to inject RestClient in lieu of CurlClient via cmake. I wanted to have both bazel and cmake builds that exercised this path. 

Ack.

> I chose a cmake install build to also make sure that the libraries were installed correctly for when we wanted to use the Rest library.

That is already tested by `cmake-install.sh`?  I do not think we need to test that twice?  Furthermore, this build does not run the integration tests, it just runs the quickstarts (see below).  I do not think you are getting the coverage you want.

Maybe something like https://github.com/googleapis/google-cloud-cpp/blob/main/ci/cloudbuild/builds/cxx20.sh would be a better starting point.  You want a build that calls this function:

https://github.com/googleapis/google-cloud-cpp/blob/3505147926c761fa457913dfb2995407077bc8cd/ci/cloudbuild/builds/lib/integration.sh#L252
___
*[`ci/cloudbuild/builds/cmake-gcs-rest-install.sh` line 184 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0l_FLY83AVQld_FIpq:-N0l_FLY83AVQld_FIpr:bvsno11) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf2da47af5dc794c3610b5ccf33790ecb635a26/ci/cloudbuild/builds/cmake-gcs-rest-install.sh#L184)):*
> ```Shell
> rm -rf "${INSTALL_PREFIX:?}"/{include,lib64}
> cmake --install cmake-out --component google_cloud_cpp_runtime
> GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT=yes quickstart::run_cmake_and_make "${INSTALL_PREFIX}"
> ```

This is only running the quickstarts, not all the integration tests.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778)* status: 12 of 16 files reviewed, 2 unresolved discussions (waiting on @coryan)

___
*[`ci/cloudbuild/builds/cmake-gcs-rest-install.sh` line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0Cajw4oV1k-B-VVSdg-r1-1:-N0lek9e4A81c9o2My0I:b10oqgz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/35cd4747ae3df21b989729f7a157098b2d120bc0/ci/cloudbuild/builds/cmake-gcs-rest-install.sh#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

> This build also defines the env var to inject RestClient in lieu of CurlClient via cmake. I wanted to have both bazel and cmake builds that exercised this path. 

Ack.

> I chose a cmake install build to also make sure that the libraries were installed correctly for when we wanted to use the Rest library.

That is already tested by `cmake-install.sh`?  I do not think we need to test that twice?  Furthermore, this build does not run the integration tests, it just runs the quickstarts (see below).  I do not think you are getting the coverage you want.

Maybe something like https://github.com/googleapis/google-cloud-cpp/blob/main/ci/cloudbuild/builds/cxx20.sh would be a better starting point.  You want a build that calls this function:

https://github.com/googleapis/google-cloud-cpp/blob/3505147926c761fa457913dfb2995407077bc8cd/ci/cloudbuild/builds/lib/integration.sh#L252

</blockquote></details>

Changed the cmake rest test to test less install and more integration, based off cxx20.
___
*[`ci/cloudbuild/builds/cmake-gcs-rest-install.sh` line 184 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0l_FLY83AVQld_FIpq:-N0lez9u3nCxl4BRbmFp:b-886gsx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcf2da47af5dc794c3610b5ccf33790ecb635a26/ci/cloudbuild/builds/cmake-gcs-rest-install.sh#L184)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This is only running the quickstarts, not all the integration tests. 
</blockquote></details>

Fixed

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-:-N0lg6u48bBmiIGfMrEY:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`ci/cloudbuild/builds/cmake-gcs-rest.sh` line 32 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0lfw8o6Mcwwn_fUptX:-N0lfw8o6Mcwwn_fUptY:b-bag9zt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1f705a0aada0936bdf604ed0ac762e9695b4a558/ci/cloudbuild/builds/cmake-gcs-rest.sh#L32)):*
> ```Shell
> env -C cmake-out GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT=yes ctest "${ctest_args[@]}" -LE "integration-test"
> 
> GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT=yes integration::ctest_with_emulators "cmake-out"
> ```

nit: seems slightly inconsistent to use `env VAR=... command` in the previous line and just `VAR=... command` in this line, your call.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`ci/cloudbuild/builds/cmake-gcs-rest.sh` line 32 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8778#-N0lfw8o6Mcwwn_fUptX:-N0lj8Kj9lzqvR3DTJP_:b-b6899l) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1f705a0aada0936bdf604ed0ac762e9695b4a558/ci/cloudbuild/builds/cmake-gcs-rest.sh#L32)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: seems slightly inconsistent to use `env VAR=... command` in the previous line and just `VAR=... command` in this line, your call.
</blockquote></details>

From what I can tell, the only reason the first invocation uses env is to change the working directory via -C which isn't necessary on the second.

**Pull request merged**